File: Same Tree/sol2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool check(TreeNode* p, TreeNode* q){
        if(p==NULL && q==NULL){
            return true;
        }
        if(q==NULL || p==NULL){
            return false;
        }
        if(p->val!=q->val){
            return false;
        }
        return true;

    }
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (p == NULL && q == NULL) {
            return true;
        }
        if(!check(p,q)){
            return false;
        }
        queue<TreeNode*> nodeQueueP;
        queue<TreeNode*> nodeQueueQ;
        nodeQueueP.push(p);
        nodeQueueQ.push(q);
        while(!nodeQueueP.empty()){
            p=nodeQueueP.front();
            q=nodeQueueQ.front();
            nodeQueueP.pop();
            nodeQueueQ.pop();

            if(!(check(p,q))){
                return false;
            }
            if(p!=NULL){
                if(!check(p->left,q->left)){
                    return false;
                }
                if(p->left!=NULL){
                    nodeQueueP.push(p->left);
                    nodeQueueQ.push(q->left);
                }

                if(!check(p->right,q->right)){
                    return false;
                }
                if(p->right!=NULL){
                    nodeQueueP.push(p->right);
                    nodeQueueQ.push(q->right);
                }

            }

        }
        return true;
    }
};

File: Balanced Binary Tree/sol2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    #bottom up recursion
    #O(n) as we only visit each node once \
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(root: Optional[TreeNode]):
            if root == None:
                return [True, 0]
            left = dfs(root.left)
            right = dfs(root.right)
            balanced = left[0] and right[0] and abs(left[1]-right[1])<=1
            return [balanced, 1+max(left[1],right[1])]
        return dfs(root)[0]